deepspeed/module_inject/replace_module.py

```diff
@@ -764,14 +781,85 @@ def replace_fn(child, _policy, layer_id=0):
+        start_time = time.time()
+        rank = dist.get_rank() if dist.is_initialized() else 0
+        world_size = dist.get_world_size() if dist.is_initialized() else 1
+        checkpoint = checkpoint_dict['checkpoints']
+        ckpt_type = checkpoint_dict['parallelization']
+        ckpt_mp_size = checkpoint_dict.get('mp_size', mp_size)
+        base_dir = checkpoint_dict.get('base_dir', None)
+
+        if ckpt_type == 'pp':
+            pbar = tqdm.tqdm(total=len(checkpoint),
+                             desc=f"Loading {len(checkpoint)} checkpoint shards")
+            for i in range(len(checkpoint)):
+                if not deepspeed.comm.is_initialized() or deepspeed.comm.get_rank() == 0:
+                    pbar.update(1)
+                sd = torch.load(checkpoint[i], map_location='cpu')
+                load_model_with_checkpoint(replaced_module, sd, mp_replace, ckpt_type)
+        else:
+            num_checkpoints = len(checkpoint) // ckpt_mp_size
+            assert world_size >= ckpt_mp_size,\
+                "Currently, merging checkpoints is not supported (when world_size is smaller than #checkpoints)!"
+            checkpoint_stride = world_size // ckpt_mp_size
+            pbar = tqdm.tqdm(total=num_checkpoints,
+                             desc=f"Loading {num_checkpoints} checkpoint shards")
+            for i in range(num_checkpoints):
+                if not deepspeed.comm.is_initialized() or deepspeed.comm.get_rank() == 0:
+                    pbar.update(1)
+
+                ckpt_index = i * ckpt_mp_size + (rank // checkpoint_stride)
+                ckpt_file = checkpoint[
+                    ckpt_index] if base_dir is None else f'{base_dir}/{checkpoint[ckpt_index]}'
+                sd = torch.load(ckpt_file, map_location='cpu')
+                load_model_with_checkpoint(replaced_module,
+                                           sd,
+                                           mp_replace,
+                                           ckpt_type,
+                                           rank % (world_size // ckpt_mp_size))
+        print(f"checkpoint loading time at rank {rank}: {time.time()-start_time} sec")
+
+    if save_mp_checkpoint_path is not None:
+        from collections import OrderedDict
+        import json
+
+        ckpt_name = checkpoint_dict['type']
+        if dist.is_initialized():
+            dist.barrier()
+        transformer_name = get_transformer_name(replaced_module)
+        non_tp_ckpt_name = f'{ckpt_name}-non-tp.pt'
+        ckpt_files = [non_tp_ckpt_name] * world_size
+        if not dist.is_initialized() or dist.get_rank() == 0:
+            print("Saving tp-sharded checkpoints")
+            torch.save(
+                OrderedDict({
+                    k: v
+                    for k,
+                    v in dict(replaced_module.state_dict()).items()
+                    if transformer_name not in k
+                }),
+                non_tp_ckpt_name)
```

Comment:
f'{save_mp_checkpoint_path}/{non_tp_ckpt_name}'

Reply:
that's true, it's not saved correctly. I am gonna fix it now

Reply:
okay, I just fixed it, you should see the following files and sizes under the save_path:
![image](https://user-images.githubusercontent.com/44502768/181653726-7fd0ae55-67c3-4975-aa13-42be979a19c7.png)

Reply:
@zcrypt0, it also generates a config file under the same path that you can use to run inference with

Reply:
https://github.com/microsoft/DeepSpeed/pull/2132#issuecomment-1198630254 is the result of using the config file after I had placed the non-tp file in the right directory

Reply:
@RezaYazdaniAminabadi Just tested and it works without a hitch, nice! :+1: